pkg/kubectl/rolling_updater.go

```diff
@@ -0,0 +1,146 @@
+	printer, err := GetPrinter("json", "", "", mapping.ObjectConvertor, nil)
+	if err != nil {
+		return nil, err
+	}
+	return &RollingUpdater{
+		Namespace:  namespace,
+		RestHelper: NewRESTHelper(client, mapping),
+		Printer:    printer,
+	}, nil
+}
+
+// Update replaces a ReplicationController (oldRc) with an updated controller
+// (newRc), by creating newRc with 0 replicas, and synchronously resizing oldRc
+// and newRc by 1 until oldRc has 0 replicas and // newRc has oldRc's original
+// # of replicas. Finally, oldRc will be deleted. If an error occurs at any
+// step of the update, it will be returned.
+func (r *RollingUpdater) Update(out io.Writer, oldRc, newRc *api.ReplicationController, interval, timeout time.Duration) error {
+	if oldRc.Spec.Replicas <= 0 {
+		return fmt.Errorf("Invalid controller spec for oldRc; required: > 0 replicas, actual: %s\n", oldRc.Spec)
+	}
+	newRc.Spec.Replicas = 0
+	oldName := oldRc.ObjectMeta.Name
+	newName := newRc.ObjectMeta.Name
+	desired := oldRc.Spec.Replicas
+
+	// push new rc
+	var buf bytes.Buffer
+	r.Printer.PrintObj(newRc, &buf)
+	fmt.Fprintf(out, "Creating %s ... ", newName)
+	err := r.RestHelper.Create(r.Namespace, true, buf.Bytes())
+	if err != nil {
+		return err
+	}
+	fmt.Fprintf(out, "Done.\n")
+
+	// refetch
+	oldRc, err = r.Get(oldName)
+	if err != nil {
+		return err
+	}
+	newRc, err = r.Get(newName)
+	if err != nil {
+		return err
+	}
+
+	// +1, -1 on oldRc, newRc until newRc has desired number of replicas
+	for desired = oldRc.Spec.Replicas; newRc.Spec.Replicas != desired; {
+		newRc.Spec.Replicas += 1
+		oldRc.Spec.Replicas -= 1
+		fmt.Fprintf(out, "Updating %s replicas: %d, %s replicas: %d ... ",
+			oldName, oldRc.Spec.Replicas,
+			newName, newRc.Spec.Replicas)
+		newRc, err = r.updateAndWait(newRc, interval, timeout)
+		if err != nil {
+			return err
+		}
+		oldRc, err = r.updateAndWait(oldRc, interval, timeout)
+		if err != nil {
+			return err
+		}
+		fmt.Fprintf(out, "Done.\n")
+	}
+	// delete old rc
+	fmt.Fprintf(out, "Update succeeded. Deleting %s ... ", oldName)
+	err = r.RestHelper.Delete(r.Namespace, oldName)
+	if err != nil {
+		return err
+	}
+	fmt.Fprintf(out, "Done.\n")
+	return nil
+}
+
+// Get returns named ReplicationController, or an error.
+func (r *RollingUpdater) Get(name string) (*api.ReplicationController, error) {
+	obj, err := r.RestHelper.Get(r.Namespace, name, labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+	return obj.(*api.ReplicationController), nil
+}
+
+func (r *RollingUpdater) updateAndWait(rc *api.ReplicationController, interval, timeout time.Duration) (*api.ReplicationController, error) {
+	var buf bytes.Buffer
+	r.Printer.PrintObj(rc, &buf)
+	name := rc.ObjectMeta.Name
+	if err := r.RestHelper.Update(r.Namespace, name, false, buf.Bytes()); err != nil {
+		return nil, err
+	}
+	if err := wait.Poll(interval, timeout, func() (bool, error) {
```

Comment:
I didn't make this change, as it's easier to test if RollingUpdater uses a client.Interface instead of the Client directly. If it's worth adding ControllerHasDesiredReplicas to the interface, I can do that.